src/provider/provider_tracking.c

```diff
@@ -37,68 +37,143 @@ struct umf_memory_tracker_t {
 typedef struct tracker_alloc_info_t {
     umf_memory_pool_handle_t pool;
     size_t size;
+    // list of previous entries with the same address (LIFO)
+    struct tracker_alloc_info_t *prev;
 } tracker_alloc_info_t;
 
-static umf_result_t umfMemoryTrackerAdd(umf_memory_tracker_handle_t hTracker,
-                                        umf_memory_pool_handle_t pool,
-                                        const void *ptr, size_t size) {
+static umf_result_t
+umfMemoryTrackerAddValue(umf_memory_tracker_handle_t hTracker, const void *ptr,
+                         tracker_alloc_info_t *new_value) {
     assert(ptr);
+    assert(new_value);
 
-    tracker_alloc_info_t *value = umf_ba_alloc(hTracker->alloc_info_allocator);
-    if (value == NULL) {
-        LOG_ERR("failed to allocate tracker value, ptr=%p, size=%zu", ptr,
-                size);
-        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
-    }
-
-    value->pool = pool;
-    value->size = size;
-
-    int ret =
-        critnib_insert(hTracker->alloc_segments_map, (uintptr_t)ptr, value, 0);
+    umf_result_t umf_result = UMF_RESULT_ERROR_UNKNOWN;
```

Comment:
Move this declaration to be sticked to this if:
```cpp
    if (ret != EEXIST) {
```
Similarly to the `int ret = ...`